x-pack/plugins/observability_ai_assistant/public/hooks/use_chat.ts

```diff
@@ -95,16 +95,41 @@ export function useChat({
 
   const handleError = useCallback(
     (error: Error) => {
-      notifications.toasts.addError(error, {
-        title: i18n.translate('xpack.observabilityAiAssistant.failedToLoadResponse', {
-          defaultMessage: 'Failed to load response from the AI Assistant',
-        }),
-      });
       if (error instanceof AbortError) {
         setChatState(ChatState.Aborted);
       } else {
         setChatState(ChatState.Error);
       }
+
+      const tokenLimitRegex =
+        /This model's maximum context length is (\d+) tokens\. However, your messages resulted in (\d+) tokens/g;
+      const tokenLimitRegexResult = tokenLimitRegex.exec(error.message);
+      if (tokenLimitRegexResult) {
+        const [, tokenLimit, tokenCount] = tokenLimitRegexResult;
+
+        setMessages((msgs) => [
+          ...msgs,
+          {
+            '@timestamp': new Date().toISOString(),
+            message: {
+              content: i18n.translate('xpack.observabilityAiAssistant.tokenLimitError', {
+                defaultMessage:
+                  'The conversation has exceeded the token limit. The maximum token limit is **{tokenLimit}**, but the current conversation has **{tokenCount}** tokens. Please start a new conversation to continue.',
+                values: { tokenLimit, tokenCount },
+              }),
+              role: MessageRole.Assistant,
+            },
+          },
+        ]);
+
+        return;
+      }
```

Comment:
Yes, there is `createTokenLimitReachedError` but it is not called when the limit is reached. I can perform the check server-side and throw the error but I'd still have to handle the error client side. Currently `createTokenLimitReachedError` invokes `ChatCompletionError` which doesn't take metadata like `tokenLimit` and  `tokenCount`. Should I add that?

Reply:
yes please!

Reply:
Not quiet working because I have trouble passing the `meta` through (working on that) but something like this?

https://github.com/elastic/kibana/pull/175871/commits/bc80bd7eb88ec0f22db564ea2e4129611a3b4c45

Reply:
yes, and maybe we can use a generic for meta? that would allow us to add a function `isTokenLimitError` that checks the code and refines to `ChatCompletionError<{ tokenLimit: number; ... }>`.

Reply:
Done